src/Kestrel.Core/Internal/Http/Http1MessageBody.cs

```diff
@@ -258,6 +258,10 @@ private void TryStopTimingReads()
 
                 if (contentLength == 0)
                 {
+                    if (upgrade && object.ReferenceEquals(context.Method, HttpMethods.Get))
+                    {
+                        return new ForUpgrade(context);
+                    }
```

Comment:
I don't love having `return new ForUpgrade(context)` in two places. I think it's simpler to just handle upgrade requests earlier in the method. How does this look instead @forki @Tratcher?

```diff
+            if (upgrade)
+            {
+                if (headers.ContentLength.HasValue && headers.ContentLength.Value != 0)
+                {
+                    context.ThrowRequestRejected(RequestRejectionReason.UpgradeRequestCannotHavePayload);
+                }
+
+                return new ForUpgrade(context);
+            }
+
             if (headers.ContentLength.HasValue)
             {
                 var contentLength = headers.ContentLength.Value;
 
                 if (contentLength == 0)
                 {
                     return keepAlive ? MessageBody.ZeroContentLengthKeepAlive : MessageBody.ZeroContentLengthClose;
                 }
-                else if (upgrade)
-                {
-                    context.ThrowRequestRejected(RequestRejectionReason.UpgradeRequestCannotHavePayload);
-                }
 
                 return new ForContentLength(keepAlive, contentLength, context);
             }
 
             // Avoid slowing down most common case
             if (!object.ReferenceEquals(context.Method, HttpMethods.Get))
             {
                 // If we got here, request contains no Content-Length or Transfer-Encoding header.
                 // Reject with 411 Length Required.
                 if (HttpMethods.IsPost(context.Method) || HttpMethods.IsPut(context.Method))
                 {
                     var requestRejectionReason = httpVersion == HttpVersion.Http11 ? RequestRejectionReason.LengthRequired : RequestRejectionReason.LengthRequiredHttp10;
                     context.ThrowRequestRejected(requestRejectionReason, context.Method);
                 }
             }
 
-            if (upgrade)
-            {
-                return new ForUpgrade(context);
-            }
-
             return keepAlive ? MessageBody.ZeroContentLengthKeepAlive : MessageBody.ZeroContentLengthClose;
         }
```

The only behavioral difference between the changes I propose above vs the change in this PR is that now POST and GET requests with a "Connection: upgrade" header will now also get upgraded instead of rejected with a 411 response. I see this as a slight improvement.

Reply:
Makes sense.